### PR TITLE
feat: add Commerce Escrow payment functionality

### LINKED ISSUE #1650: Checkout Commerce Wrapper Smart Contract
## Problem 

We need to implement the Wrapper Smart Contract around the Base Commerce Escrow Contract

### PR DESCRIPTION
**Pull request by rodrigopavezi**
- Introduced Commerce Escrow payment types and parameters in the payment-types module.
- Added Commerce Escrow wrapper to smart contracts and updated related scripts for deployment and verification.
- Updated Docker Compose file to specify platform for services.
- Added commerce-payments dependency in smart contracts package.

## Summary by CodeRabbit

* **New Features**
  * Added ERC20 Commerce Escrow Wrapper for managing authorized payments with capture, void, charge, reclaim, and refund capabilities.
  * Added Base Sepolia testnet support with USDC token integration.

* **Documentation**
  * Added fee mechanism design documentation for Commerce Escrow payments.

* **Tests**
  * Added comprehensive test suites for Commerce Escrow Wrapper functionality.
  * Enhanced security testing with automated fuzzing and static analysis.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

Fix #1650

### CODE REVIEW

**Comment by rodrigopavezi** [comment]
@coderabbitai review

File: packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+    // Create PaymentInfo for the capture operation (must match the original authorization)
+    IAuthCaptureEscrow.PaymentInfo memory paymentInfo = _createPaymentInfoFromStored(
+      payment,
+      paymentReference
+    );
+
+    // Capture from escrow with NO FEE - let ERC20FeeProxy handle fee distribution
+    // This way the wrapper receives the full captureAmount
+    commerceEscrow.capture(paymentInfo, captureAmount, 0, address(0));
+
```
Comment (MantisClone) [inline packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol]: **Commerce Escrow already prevents double-capture**

The wrapper never stores `capturableAmount` - it always queries `commerceEscrow.paymentState()`. Commerce Escrow already prevents double-capture by tracking `capturableAmount` and reverting with `InsufficientAuthorization` if capture exceeds available amount.

No need to add wrapper-level state tracking for payment finalization - delegate state management to Commerce Escrow.

File: packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+  IAuthCaptureEscrow public immutable commerceEscrow;
+
+  /// @notice Request Network's ERC20FeeProxy contract
+  IERC20FeeProxy public immutable erc20FeeProxy;
+
+  /// @notice Maps Request Network payment references to internal payment data
+  mapping(bytes8 => PaymentData) public payments;
+
+  /// @notice Internal payment data structure
+  struct PaymentData {
```
Comment (MantisClone) [inline packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol]: **Consider delegating existence check to Commerce Escrow**

The `isActive` boolean duplicates information already tracked by Commerce Escrow's `paymentState[hash].hasCollectedPayment`. 

Could simplify by checking `commerceEscrow.paymentState(hash).hasCollectedPayment` instead of maintaining separate wrapper state. This would eliminate potential state synchronization issues and reduce storage costs.

File: packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+    // Capture from escrow with NO FEE - let ERC20FeeProxy handle fee distribution
+    // This way the wrapper receives the full captureAmount
+    commerceEscrow.capture(paymentInfo, captureAmount, 0, address(0));
+
+    // Calculate fee amounts - ERC20FeeProxy will handle the split
+    uint256 feeAmount = (captureAmount * feeBps) / 10000;
+    uint256 merchantAmount = captureAmount - feeAmount;
+
+    // Approve ERC20FeeProxy to spend the full amount we received
+    IERC20(payment.token).forceApprove(address(erc20FeeProxy), captureAmount);
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol]: `forceApprove` is required for **USDT compatibility**. 

USDT requires allowances to be reset to 0 before setting a new non-zero value. Regular `approve()` will revert if there's any existing allowance.

`forceApprove` (from OpenZeppelin's SafeERC20) handles this automatically:
1. Tries `approve(spender, value)`
2. If that fails, does `approve(spender, 0)` then `approve(spender, value)`

This is the recommended OpenZeppelin pattern for supporting USDT and other non-standard tokens.

**Review by MantisClone — COMMENTED** [review]
**Partial review - ERC20CommerceEscrowWrapper.sol only**

I've reviewed ERC20CommerceEscrowWrapper.sol and posted comments on:
- Missing `feeBps` upper bound validation (high priority)
- `PaymentData` struct packing opportunity (45% storage savings)
- `InvalidPayer` error for better debugging
- State delegation to Commerce Escrow (architectural note)

Haven't reviewed the other 16 files yet. The contract looks solid overall - main concerns are the fee validation and storage optimization.

File: packages/smart-contracts/test/contracts/ERC20CommerceEscrowWrapper.test.ts [hunk 4]
```diff
@@ -0,0 +1,1109 @@
...
+        refundExpiry,
+        feeBps,
+        feeReceiver: feeReceiverAddress,
+        tokenCollector: tokenCollectorAddress,
+        collectorData: '0x',
+      };
+
+      await expect(wrapper.chargePayment(chargeParams)).to.emit(wrapper, 'PaymentCharged');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/smart-contracts/test/contracts/ERC20CommerceEscrowWrapper.test.ts]: **Confirming this finding.** 

The "Reentrancy Protection" test suite (lines 816-883) validates that operations succeed under the `nonReentrant` modifier, but doesn't actually test reentrancy attack scenarios.

**What's missing:**
A malicious ERC20 contract that attempts reentrant calls during token transfer callbacks. For example:

```solidity
contract MaliciousReentrantToken is ERC20 {
  ERC20CommerceEscrowWrapper public target;
  
  function transfer(address to, uint256 amount) public override returns (bool) {
    // Attempt reentrancy during transfer
    if (to == address(target)) {
      target.capturePayment(...); // Should revert with ReentrancyGuard
    }
    return super.transfer(to, amount);
  }
}
```

**Recommendation:** Either implement proper reentrancy tests with malicious contracts, or remove these tests and document that reentrancy protection is verified through code review and the `nonReentrant` modifier from OpenZeppelin's battle-tested ReentrancyGuard.

File: packages/smart-contracts/test/contracts/ERC20CommerceEscrowWrapper.test.ts [hunk 5]
```diff
@@ -0,0 +1,1109 @@
...
+        refundExpiry,
+        feeBps,
+        feeReceiver: feeReceiverAddress,
+        tokenCollector: tokenCollectorAddress,
+        collectorData: '0x',
+      };
+    });
+
+    it('should charge payment successfully', async () => {
+      const expectedFeeAmount = amount.mul(feeBps).div(10000);
```
Comment (MantisClone) [inline packages/smart-contracts/test/contracts/ERC20CommerceEscrowWrapper.test.ts]: **Fee calculation tests verify events but not actual token amounts**

The test calculates `expectedFeeAmount` but only verifies event emissions - it never checks that the fee receiver's token balance actually increased by this amount.

**Current gap:**
```typescript
const expectedFeeAmount = amount.mul(feeBps).div(10000);
// ... test only checks events, not balanceOf()
```

**Recommended improvement to existing fee tests:**
Add balance verification to existing `chargePayment` tests (and similar capture tests):

```typescript
// Before charging payment
const merchantBalanceBefore = await testERC20.balanceOf(merchantAddress);
const feeReceiverBalanceBefore = await testERC20.balanceOf(feeReceiverAddress);

const tx = await wrapper.chargePayment(chargeParams);

// After charging payment - verify actual token transfers
const merchantBalanceAfter = await testERC20.balanceOf(merchantAddress);
const feeReceiverBalanceAfter = await testERC20.balanceOf(feeReceiverAddress);

const expectedFeeAmount = amount.mul(feeBps).div(10000);
const expectedMerchantAmount = amount.sub(expectedFeeAmount);

expect(merchantBalanceAfter.sub(merchantBalanceBefore)).to.equal(expectedMerchantAmount);
expect(feeReceiverBalanceAfter.sub(feeReceiverBalanceBefore)).to.equal(expectedFeeAmount);
```

**Also test edge cases:**
- 0% fee (feeBps = 0): verify merchant gets full amount, fee receiver gets nothing
- 100% fee (feeBps = 10000): verify merchant gets nothing, fee receiver gets all
- 2.5% fee (current): verify exact split matches calculation

This ensures fees are actually transferred, not just emitted in events.

**Severity:** High - Missing critical validation of actual token transfers

File: packages/payment-processor/src/payment/erc20-commerce-escrow-wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+  provider,
+}: {
+  params: AuthorizePaymentParams;
+  network: CurrencyTypes.EvmChainName;
+  provider: providers.Provider | Signer;
+}): string {
+  const wrapperContract = erc20CommerceEscrowWrapperArtifact.connect(network, provider);
+
+  // Use utils.Interface to encode with the raw ABI to avoid TypeScript interface issues
+  const iface = new utils.Interface(
```
Comment (MantisClone) [inline packages/payment-processor/src/payment/erc20-commerce-escrow-wrapper.ts]: **Inconsistent Interface usage pattern**

`encodeAuthorizePayment` and `encodeChargePayment` use `utils.Interface` workaround (lines 127, 218) while `encodeCapturePayment` uses direct `interface.encodeFunctionData` (line 169).

**Reason:** Functions with many parameters (12+ params) hit TypeScript limitations when encoding. The workaround is needed for authorize/charge but not for simpler functions like capture (4 params).

**Recommendation:** Add JSDoc comment to document the pattern:

```typescript
/**
 * @remarks
 * Uses utils.Interface to handle large parameter count (TypeScript encoding limitation).
 * Similar pattern in single-request-forwarder.ts.
 */
```

This makes the inconsistency intentional rather than appearing accidental.

File: packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/smart-contracts/src/contracts/ERC20CommerceEscrowWrapper.sol]: done

File: packages/smart-contracts/src/contracts/test/EchidnaERC20CommerceEscrowWrapper.sol
Comment (MantisClone) [inline packages/smart-contracts/src/contracts/test/EchidnaERC20CommerceEscrowWrapper.sol]: ## Echidna Tests Not Running?

Hey @rodrigopavezi! Quick observation: the Echidna CI runs all show `Passed: 0, Failed: 0, Total: 0, Pass Rate: 0%`, which suggests Echidna isn't finding any properties to test.

The harness has 16 `echidna_*` invariant functions that look reasonable for fuzzing:

**Good fuzz coverage:**
- Fee calculation bounds with random amounts (prevents overcharging edge cases)
- Token conservation across random operation sequences
- Accounting invariants with fuzzed amounts (captured ≤ authorized, etc.)
- Overflow/underflow checks with extreme values

**Possibly redundant:**
- `echidna_merchant_receives_nonnegative` - uint256 can't be negative anyway
- `echidna_payment_ref_counter_monotonic` - uint256 ≥ 0 always true
- `echidna_fee_bps_validation_enforced` - Just returns `true`

**Main concern:** The CI shows 0 tests run, which means either:
1. Echidna isn't discovering the property functions, or
2. The `driver_*` functions aren't being called to create fuzzable state transitions

The driver functions (`driver_authorizePayment`, `driver_capturePayment`, etc.) are what should let Echidna explore random sequences of operations and find edge cases. Without those running, the invariants have nothing to check.

Could you take a look when you have a chance? Not sure if it's a config issue or something else.

**Debug idea**: Running `echidna . --contract EchidnaERC20CommerceEscrowWrapper --config echidna.config.yml --format text` locally might show what Echidna is discovering.

File: packages/smart-contracts/ECHIDNA_CHANGES_SUMMARY.md
Comment (MantisClone) [inline packages/smart-contracts/ECHIDNA_CHANGES_SUMMARY.md]: @rodrigopavezi was the addition of this file intentional?  I think we should remove it.
